File: LeetCode/합집합 찾기 3 제한된 친구 요청 처리.py
class Solution:
    def friendRequests(self, n: int, restrictions: List[List[int]], requests: List[List[int]]) -> List[bool]:
        # 제한 관계를 O(1)로 빠르게 조회하기 위해 set으로 저장
        rests = set([tuple(r) for r in restrictions])
        # 각 사람이 속한 친구 그룹을 저장하는 딕셔너리
        sets = {}
        results = []

        # frozenset: 변경 불가능한 속성이 추가된 set(Read only)
        # 딕셔너리 value로 안전하게 공유가 가능하고 union 결과를 여러 key에 동일 객체로 할당 가능
        for request in requests:
            # 현재 두 사람의 소속 집합 탐색
            # 이미 그룹이 있으면 그 그룹, 없으면 자기 자신만 포함된 집합
            from_set = sets.get(request[0], frozenset([request[0]]))
            to_set = sets.get(request[1], frozenset([request[1]]))

            # 제한 위반 검사: from_set과 to_set의 쌍 중 하나라도 rests에 있는지 검사
            if any((x, y) in rests or (y, x) in rests for x, y in itertools.product(from_set, to_set)):
                results.append(False)
            else:
                # 위반이 없으면 Union 수행
                # 모든 구성원에게 같은 집합을 매핑
                results.append(True)
                union = from_set.union(to_set)
                for num in union:
                    sets[num] = union
        
        return results

File: LeetCode/합집합 찾기 3 제한된 친구 요청 처리.cpp
class Solution {
public:
    vector<bool> friendRequests(int n, vector<vector<int>>& restrictions, vector<vector<int>>& requests) {
        // union-find를 위한 초기 부모 배열
        // 각 인원은 자기 자신의 부모
        vector<int> parent(n);
        // iota: 순차적인 순번을 부여하여 초기화
        iota(parent.begin(), parent.end(), 0);

        // 경로 압축이 있는 함수 탐색
        // 주어진 노드의 루트 부모를 반환
        function<int(int)> find = [&](int x) {
            if (parent[x] != x) {
                // 경로 압축: 노드가 바로 부모를 볼 수 있도록 함
                parent[x] = find(parent[x]);
            }
            return parent[x];
        };

        // 각 요청이 처리 가능한지 확인하는 결과 벡터
        vector<bool> result;

        // 각 친구 요청 처리
        for (auto& request : requests) {
            int person1 = request[0];
            int person2 = request[1];

            // 두 사람의 부모 루트 탐색
            int root1 = find(person1);
            int root2 = find(person2);

            // 이미 같은 그룹에 있다면 같은 친구 그룹에 이미 속하므로 자동으로 처리
            if (root1 == root2) {
                result.push_back(true);
            // 같은 그룹에 없다면
            } else {
                // 두 그룹을 합치는게 규정 위반인지 체크
                bool canMerge = true;

                for (auto& restriction : restrictions) {
                    // 각 제약 쌍을 가져오기
                    int restRoot1 = find(restriction[0]);
                    int restRoot2 = find(restriction[1]);

                    // 해당 합병이 제약 있는 쌍을 연결하는지 체크
                    // 각 제약 있는 인원이 각각의 그룹에 속할 때 불가 
                    if ((root1 == restRoot1 && root2 == restRoot2) || 
                        (root2 == restRoot1 && root1 == restRoot2)) {
                        canMerge = false;
                        break;
                    }
                }
                result.push_back(canMerge);

                // 합병 가능한 경우 두 그룹을 합치기
                if (canMerge) {
                    // Union 작업
                    parent[root1] = root2;
                }
            }
        }
        return result;
    }
};